src/foam/nanos/menu/Menu.js

```diff
@@ -70,13 +82,77 @@ foam.CLASS({
       name: 'icon',
       documentation: 'Icon associated to the menu item.',
       displayWidth: 80
+    },
+    {
+      documentation: 'Predicate providing arbitrary checks, in addition to the regular menu auth checks.',
+      class: 'foam.mlang.predicate.PredicateProperty',
+      name: 'readPredicate',
+      view: {
+        class: 'foam.u2.view.JSONTextView'
+      },
+      javaFactory: `
+        return foam.mlang.MLang.TRUE;
+      `,
     }
   ],
 
   methods: [
     function launch_(X, e) {
       this.lastMenuLaunchedListener && this.lastMenuLaunchedListener(this);
       this.handler && this.handler.launch(X, this, e);
+    },
+    {
+      documentation: 'Desire to call read predicate with calling context but predicate may also need access to this menu; add the current menu as context key MENU',
+      name: 'f',
+      type: 'Boolean',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        }
+      ],
+      javaCode: `
+        return getReadPredicate().f(
+          x.put("MENU", this)
+        );
+      `
+    },
+    {
+      name: 'authorizeOnCreate',
+      javaCode: `
+        AuthService auth = (AuthService) x.get("auth");
+        if ( ! auth.check(x, "menu.create") ) {
+          throw new AuthorizationException("You do not have permission to create menus.");
+        }
+      `
+    },
+    {
+      name: 'authorizeOnUpdate',
+      javaCode: `
+        AuthService auth = (AuthService) x.get("auth");
+        if ( ! auth.check(x, "menu.update." + getId()) ) {
+          throw new AuthorizationException("You do not have permission to update this menu.");
+        }
+      `
+    },
+    {
+      name: 'authorizeOnDelete',
+      javaCode: `
+        AuthService auth = (AuthService) x.get("auth");
+        if ( ! auth.check(x, "menu.remove." + getId()) ) {
+          throw new AuthorizationException("You do not have permission to delete menus.");
+        }
+      `
+    },
+    {
+      name: 'authorizeOnRead',
+      javaCode: `
+        AuthService auth = (AuthService) x.get("auth");
+        if ( ! ( auth.check(x, "menu.read." + getId()) &&
```

Comment:
Potential optimization: if we switch the order so that it checks the predicate first, then we can short circuit and avoid the relatively expensive permission check when the predicate is false. Chances are that the predicate is almost always going to be cheaper than the permission check.

```Java
! ( false && x ) // always true, don't need to evaluate x
```

Reply:
Incorporated